tests_ci/dummy_image.py

```diff
@@ -0,0 +1,247 @@
+
+
+def output_writer(f_in, f_out, size, db=None):
+    if not f_in:
+        util.zero_n(f_out, size)
+    else:
+        crc_hash = None
+        mapping = None
+
+        if db is not None and size != 0:
+            mapping = True
+
+        if mapping:
+            crc_hash = Crc32Hasher()
+
+            # Combine sections if adjacent
+            if db and db[-1]["range"][1] == f_out.tell() - 1:
+                entry = db.pop()
+                crc_hash.set_checksum(entry["checksum"])
+                section_start = entry["range"][0]
+            else:
+                section_start = f_out.tell()
+
+            section_end = f_out.tell() + (size - 1)
+
+        util.copyfileobj_n(f_in, f_out, size, hasher=crc_hash)
+
+        if mapping:
+            db.append(
+                {
+                    "range": [section_start, section_end],
+                    "checksum": crc_hash.get_checksum(),
+                }
+            )
+
+
+def download_file(f_out, url, section):
+    crc_hash = Crc32Hasher()
+
+    req = urllib.request.Request(url)
+    req.add_header("Range", f"bytes={section['range'][0]}-{section['range'][1]}")
+    with urllib.request.urlopen(req) as data:
+        util.copyfileobj_n(
+            data, f_out, section["range"][1] - section["range"][0] + 1, hasher=crc_hash
+        )
+
+    if crc_hash.get_checksum() != int(section["checksum"], 16):
+        raise Exception(f"ERROR: Checksum of range {section['range']} doesn't match")
+
+
+def download_dummy_image(args):
+    with open(args.db, "r") as f:
+        database = f.read()
+
+    database = tomlkit.parse(database)
+
+    device_entry = database["device"].get(args.device_name)
+    if not device_entry:
+        raise Exception(f"ERROR: {args.device_name} cannot be found in the database")
+
+    # Assume we always download the beginning and end (which makes sense for a ZIP file)
+    previous_offset = 0
+    with output_open(args.output, "wb", args.compress) as f_out:
+        for section in device_entry["sections"]:
+            util.zero_n(f_out, section["range"][0] - previous_offset)
+            download_file(f_out, device_entry["url"], section)
+            previous_offset = section["range"][1] + 1
+
+
+def print_entry(db, original_filename):
+    new_map = tomlkit.array()
+    new_map.append(tomlkit.nl())
+
+    for section in db:
+        entry = tomlkit.inline_table()
+
+        for key, value in section.items():
+            entry[key] = f"{value:x}" if key == "checksum" else value
+
+        new_map.append(entry)
+
+    print("*** Add this entry to the database ***")
+    print(
+        tomlkit.dumps(
+            {
+                "device": {
+                    "new_device_name": {
+                        "filename": os.path.basename(original_filename),
+                        "url": "",
+                        "sections": new_map,
+                    }
+                }
+            }
+        )
+    )
+    print("*** End ***")
```

Comment:
Yes I have been struggling with creating TOML files with that specific inline table structure. I like that format to keep the file compact instead of creating a new table for each section. I'll see if it will correctly preserve the structure.

Reply:
Ah, fair enough. I'm totally fine with a manual copy/paste if the library doesn't have a way to produce a table with decent formatting.